File: srcs/server/orchestration/queue/postgres_queue.go
package queue

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/onehumancorp/mono/srcs/server/db"
	"github.com/onehumancorp/mono/srcs/server/telemetry"
)

type PostgresTaskQueue struct {
	provider db.Provider
}

func NewPostgresTaskQueue(provider db.Provider) *PostgresTaskQueue {
	return &PostgresTaskQueue{provider: provider}
}

func (q *PostgresTaskQueue) Enqueue(ctx context.Context, job *Job) error {
	defer func() {
		telemetry.RecordQueueLength(ctx, 1) // Approximation
	}()

	if job.RunAfter.IsZero() {
		job.RunAfter = time.Now()
	}

	// Make sure payload contains the agent_role
	var payloadMap map[string]interface{}
	if err := json.Unmarshal([]byte(job.Payload), &payloadMap); err != nil {
		payloadMap = map[string]interface{}{}
	}
	payloadMap["agent_role"] = job.AgentRole
	payloadMap["attempts"] = job.Attempts
	payloadMap["max_attempts"] = job.MaxAttempts

	newPayload, _ := json.Marshal(payloadMap)

	orgID := ""
	if val, ok := payloadMap["organization_id"].(string); ok {
		orgID = val
	}

	query := `
		INSERT INTO sub_agent_queue (
			id, organization_id, parent_task_id, payload, status, scheduled_at
		) VALUES (
			$1, $2, $3, $4, $5, $6
		)
	`
	_, err := q.provider.Exec(ctx, query,
		job.ID, orgID, job.ParentTaskID, string(newPayload),
		"PENDING", job.RunAfter,
	)
	return err
}

func (q *PostgresTaskQueue) Dequeue(ctx context.Context, roles []string) (*Job, error) {
	// Select a PENDING job. We use FOR UPDATE SKIP LOCKED
	// Agent role is in the payload. We extract it using JSON ops ->> 'agent_role'

	rolePlaceholders := make([]string, len(roles))
	args := []any{}

	for i, role := range roles {
		rolePlaceholders[i] = fmt.Sprintf("$%d", i+1)
		args = append(args, role)
	}

	roleFilter := ""
	if len(roles) > 0 {
		roleFilter = fmt.Sprintf("AND payload::json->>'agent_role' IN (%s)", strings.Join(rolePlaceholders, ", "))
	}

	now := time.Now()
	args = append(args, now)
	nowIdx := len(args)

	query := fmt.Sprintf(`
		UPDATE sub_agent_queue
		SET status = 'RUNNING'
		WHERE id = (
			SELECT id
			FROM sub_agent_queue
			WHERE status = 'PENDING'
			  AND scheduled_at <= $%d
			  %s
			ORDER BY scheduled_at ASC
			FOR UPDATE SKIP LOCKED
			LIMIT 1
		)
		RETURNING id, organization_id, parent_task_id, payload, status, scheduled_at
	`, nowIdx, roleFilter)

	var j Job
	var scheduledAt time.Time
	var orgID string

	err := q.provider.QueryRow(ctx, query, args...).Scan(
		&j.ID, &orgID, &j.ParentTaskID, &j.Payload, &j.Status, &scheduledAt,
	)

	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil // No jobs available
	} else if err != nil {
		return nil, err
	}

	telemetry.RecordQueueLength(ctx, -1)
		telemetry.RecordSubAgentQueueDelay(ctx, time.Since(scheduledAt).Seconds())

	j.RunAfter = scheduledAt

	// Reconstruct other fields from payload
	var payloadMap map[string]interface{}
	if err := json.Unmarshal([]byte(j.Payload), &payloadMap); err == nil {
		if role, ok := payloadMap["agent_role"].(string); ok {
			j.AgentRole = role
		}
		if attempts, ok := payloadMap["attempts"].(float64); ok {
			j.Attempts = int(attempts)
		}
		if maxAttempts, ok := payloadMap["max_attempts"].(float64); ok {
			j.MaxAttempts = int(maxAttempts)
		}
	}

	j.Attempts++

	return &j, nil
}

func (q *PostgresTaskQueue) Complete(ctx context.Context, jobID string) error {
	query := `
		UPDATE sub_agent_queue
		SET status = 'COMPLETED', completed_at = $1
		WHERE id = $2
	`
	_, err := q.provider.Exec(ctx, query, time.Now(), jobID)
	return err
}

func (q *PostgresTaskQueue) Fail(ctx context.Context, jobID string, reason string) error {
	// First fetch the payload
	query := `SELECT payload FROM sub_agent_queue WHERE id = $1`
	var payloadStr string
	err := q.provider.QueryRow(ctx, query, jobID).Scan(&payloadStr)
	if err != nil {
		return err
	}

	var payload map[string]interface{}
	json.Unmarshal([]byte(payloadStr), &payload)
	if payload == nil {
		payload = make(map[string]interface{})
	}

	attempts := 1
	if a, ok := payload["attempts"].(float64); ok {
		attempts = int(a)
	}

	maxAttempts := 3
	if m, ok := payload["max_attempts"].(float64); ok {
		maxAttempts = int(m)
	}

	var status string
	var nextRunAfter time.Time

	if attempts >= maxAttempts {
		status = "FAILED"
		nextRunAfter = time.Now()
	} else {
		status = "PENDING"
		backoff := time.Duration(1<<attempts) * time.Second
		nextRunAfter = time.Now().Add(backoff)
		telemetry.RecordQueueLength(ctx, 1) // Job returned to queue
	}

	payload["last_error"] = reason
	newPayload, _ := json.Marshal(payload)

	updateQuery := `
		UPDATE sub_agent_queue
		SET status = $1, scheduled_at = $2, payload = $3
		WHERE id = $4
	`
	_, err = q.provider.Exec(ctx, updateQuery, status, nextRunAfter, string(newPayload), jobID)
	return err
}

File: src/server/orchestration/queue/postgres_queue.rs
use super::queue::{Job, TaskQueue};
use async_trait::async_trait;
use sqlx::{PgPool, Row};
use std::sync::Arc;

pub struct PostgresTaskQueue {
    pool: Arc<PgPool>,
}

impl PostgresTaskQueue {
    pub fn new(pool: Arc<PgPool>) -> Self {
        Self { pool }
    }
}

#[async_trait]
impl TaskQueue for PostgresTaskQueue {
    async fn enqueue_batch(&self, jobs: Vec<Job>) -> Result<(), String> {
        if jobs.is_empty() { return Ok(()); }
        let mut tx = self.pool.begin().await.map_err(|e| e.to_string())?;
        for job in jobs {
            sqlx::query(
                "INSERT INTO sub_agent_jobs (id, organization_id, parent_task_id, agent_role, payload, status, run_after)
                 VALUES ($1, $2, $3, $4, $5, 'QUEUED', $6)"
            )
            .bind(&job.id)
            .bind(&job.tenant_id)
            .bind(&job.parent_task_id)
            .bind(&job.agent_role)
            .bind(serde_json::from_str::<serde_json::Value>(&job.payload).unwrap_or_else(|_| serde_json::json!({})))
            .bind(job.run_after)
            .execute(&mut *tx)
            .await
            .map_err(|e| e.to_string())?;
        }
        tx.commit().await.map_err(|e| e.to_string())?;
        Ok(())
    }

    async fn enqueue(&self, job: Job) -> Result<(), String> {
        sqlx::query(
            "INSERT INTO sub_agent_jobs (id, organization_id, parent_task_id, agent_role, payload, status, run_after)
             VALUES ($1, $2, $3, $4, $5, 'QUEUED', $6)"
        )
        .bind(&job.id)
        .bind(&job.tenant_id)
        .bind(&job.parent_task_id)
        .bind(&job.agent_role)
        .bind(serde_json::from_str::<serde_json::Value>(&job.payload).unwrap_or_else(|_| serde_json::json!({})))
        .bind(job.run_after)
        .execute(&*self.pool)
        .await
        .map_err(|e| e.to_string())?;

        Ok(())
    }

    async fn dequeue(&self, roles: Vec<String>) -> Result<Option<Job>, String> {
        if roles.is_empty() {
            return Ok(None);
        }

        let mut tx = self.pool.begin().await.map_err(|e| e.to_string())?;

        // Bypass RLS for queue operations to avoid issues where the queue processor
        // might not have the correct tenant context.
        sqlx::query("SET LOCAL ROLE ohc_bypassrls")
            .execute(&mut *tx)
            .await
            .map_err(|e| e.to_string())?;

        let row_opt = sqlx::query(
            "UPDATE sub_agent_jobs
             SET status = 'RUNNING', updated_at = CURRENT_TIMESTAMP
             WHERE id = (
                 SELECT id
                 FROM sub_agent_jobs
                 WHERE status = 'QUEUED'
                   AND agent_role = ANY($1)
                   AND run_after <= CURRENT_TIMESTAMP
                 ORDER BY run_after ASC
                 LIMIT 1
                 FOR UPDATE SKIP LOCKED
             )
             RETURNING id, organization_id, parent_task_id, agent_role, payload, status, attempts, max_attempts, run_after, locked_until, created_at, updated_at"
        )
        .bind(&roles)
        .fetch_optional(&mut *tx)
        .await
        .map_err(|e| e.to_string())?;

        if let Some(row) = row_opt {
            let payload: serde_json::Value = row.get("payload");
            let job = Job {
                id: row.get("id"),
                tenant_id: row.try_get("organization_id").unwrap_or_default(),
                parent_task_id: row.get("parent_task_id"),
                agent_role: row.get("agent_role"),
                payload: serde_json::to_string(&payload).unwrap_or_default(),
                status: row.get("status"),
                attempts: row.get("attempts"),
                max_attempts: row.get("max_attempts"),
                run_after: row.try_get("run_after").unwrap_or_else(|_| chrono::Utc::now()),
                locked_until: row.try_get("locked_until").unwrap_or(None),
                created_at: row.try_get("created_at").unwrap_or_else(|_| chrono::Utc::now()),
                updated_at: row.try_get("updated_at").unwrap_or_else(|_| chrono::Utc::now()),
            };

            tx.commit().await.map_err(|e| e.to_string())?;
            return Ok(Some(job));
        }

        tx.commit().await.map_err(|e| e.to_string())?;
        Ok(None)
    }

    async fn complete(&self, job_id: &str) -> Result<(), String> {
        let mut tx = self.pool.begin().await.map_err(|e| e.to_string())?;
        sqlx::query("SET LOCAL ROLE ohc_bypassrls")
            .execute(&mut *tx)
            .await
            .map_err(|e| e.to_string())?;

        sqlx::query("UPDATE sub_agent_jobs SET status = 'COMPLETED', updated_at = CURRENT_TIMESTAMP WHERE id = $1")
            .bind(job_id)
            .execute(&mut *tx)
            .await
            .map_err(|e| e.to_string())?;

        tx.commit().await.map_err(|e| e.to_string())?;
        Ok(())
    }

    async fn fail(&self, job_id: &str, _reason: &str) -> Result<(), String> {
        let mut tx = self.pool.begin().await.map_err(|e| e.to_string())?;
        sqlx::query("SET LOCAL ROLE ohc_bypassrls")
            .execute(&mut *tx)
            .await
            .map_err(|e| e.to_string())?;

        sqlx::query("UPDATE sub_agent_jobs SET status = 'FAILED', updated_at = CURRENT_TIMESTAMP WHERE id = $1")
            .bind(job_id)
            .execute(&mut *tx)
            .await
            .map_err(|e| e.to_string())?;

        tx.commit().await.map_err(|e| e.to_string())?;
        Ok(())
    }
}
